### ISSUE TITLE
Object `__call__` decorated with in `lru_cache` is not recognized

### ISSUE THREAD

**Issue by CaselIT** [microsoft/pyright#4015]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**

Defining a class with a `__call__` method decorated with an `lru_cache` is not recognized by pyright

**To Reproduce**
Steps to reproduce the behavior.

run pyright on the code below. the output is
```
try.py
  try.py:20:13 - information: Type of "Y()(1)" is "int"
  try.py:21:13 - error: Object of type "X" is not callable (reportGeneralTypeIssues)
  try.py:21:13 - information: Type of "X()(1)" is "Unknown"
  try.py:22:13 - information: Type of "X().foo(1)" is "int"
```

**Expected behavior**
mypy will correctly report
```
try.py:20: note: Revealed type is "builtins.int"
try.py:21: note: Revealed type is "builtins.int"
try.py:22: note: Revealed type is "builtins.int"
Success: no issues found in 1 source file
```
and python will print
```
3
3
3
```

without lru_cache pyright works correctly

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from functools import lru_cache
from typing import TYPE_CHECKING

class Y:
    def __call__(self, v: int) -> int:
        return v + 2

class X:
    @lru_cache
    def __call__(self, v: int) -> int:
        return v + 2

    @lru_cache
    def foo(self, v: int) -> int:
        return v + 2

if not TYPE_CHECKING:
    reveal_type = print

reveal_type(Y()(1))
reveal_type(X()(1))
reveal_type(X().foo(1))
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

I've noticed this when using pylance on vscode, but the above output comes from running pyright directly in the terminal

**Additional context**

pyright 1.1.274
pylance: Pylance language server 2022.9.40

**Comment by erictraut** [comment]
Interesting. I didn't realize that the Python runtime allowed chaining of calls through multiple objects that support the `__call__` protocol. For example, the following works fine.

```python
class A:
    def __call__(self, v: int):
        print("Received", v)

class B:
    __call__ = A()

class C:
    __call__ = B()

class D:
    __call__ = C()

d = D()

d(1)
```

I've updated pyright's logic to handle this case, which fixes the reported problem. This will be addressed in the next release.

**Comment by CaselIT** [comment]
Thanks for the fix

I had no idea you could do the above, but I guess from python point of view a function is just an object with a `__call__` method

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.275, which I just published. It will also be included in a future release of pylance.
